### PR TITLE
Add support for analytic solution of inhomogeneous ODEs

### PR DESCRIPTION
**Pull request by clinssen**

### CODE REVIEW

File: odetoolbox/shapes.py [hunk 1]
```diff
@@ -37,7 +41,15 @@ class MalformedInputException(Exception):
     pass
 
 
-class Shape():
+def is_constant_term(term, parameters: Mapping[sympy.Symbol, str]=None):
```
Comment (pnbabu) [inline odetoolbox/shapes.py:44]: Could you add docstrings?

File: odetoolbox/system_of_shapes.py [hunk 2]
```diff
@@ -179,9 +188,21 @@ def generate_propagator_solver(self, output_timestep_symbol="__h"):
...
+                        # check that off-diagonal elements are zero in the system matrix to check that this is at most a first-order system
+                        for _col in range(P_sym.shape[1]):
+                            if _col != row:
+                                assert _is_zero(self.A_[row, _col]), "Higher-order inhomogeneous ODEs are not supported"
+
+                        particular_solution = -self.b_[row] / self.A_[row, row]
+                        update_expr_terms.append(sym_str + " * (" + str(self.x_[col]) + " - (" + str(particular_solution) + "))" + " + (" + str(particular_solution) + ")")
+                    assert _is_zero(self.c_[row]), "nonlinear part should be zero for propagators"
+            update_expr[str(self.x_[row])] = " + ".join(update_expr_terms)
+            update_expr[str(self.x_[row])] = sympy.parsing.sympy_parser.parse_expr(update_expr[str(self.x_[row])], global_dict=Shape._sympy_globals)
```
Comment (pnbabu) [inline odetoolbox/system_of_shapes.py:205]: Why is the `sympy.simplify` removed here?

File: tests/test_lin_const_coeff_and_homogeneous.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        shape_V_m_lin = Shape.from_ode("V_m", "-V_m/Tau + (I_in + I_ex + I_e) / C_m", initial_values={"V_m": "0."}, parameters=self._parameters)
+        shape_V_m_lin_no_param = Shape.from_ode("V_m", "-V_m/Tau + (I_in + I_ex + I_e) / C_m", initial_values={"V_m": "0."})
+        shape_V_m_nonlin = Shape.from_ode("V_m", "-V_m**2/Tau + (I_in + I_ex + I_e) / C_m", initial_values={"V_m": "0."}, parameters=self._parameters)
+        shape_V_m_nonlin_no_param = Shape.from_ode("V_m", "-V_m**2/Tau + (I_in + I_ex + I_e) / C_m", initial_values={"V_m": "0."})
+
+        for shape in [shape_inh, shape_exc]:
+            self.assertTrue(shape.is_lin_const_coeff())
+            self.assertTrue(shape.is_homogeneous())
+        self.assertTrue(shape.is_lin_const_coeff())
+        self.assertTrue(shape.is_homogeneous())
```
Comment (pnbabu) [inline tests/test_lin_const_coeff_and_homogeneous.py:106]: Are these lines required?

File: odetoolbox/system_of_shapes.py (hunk 2 above)
Comment (clinssen) [reply odetoolbox/system_of_shapes.py:205]: It's a summation of terms of the form c1*x + c2*x' + c3*x'', so there's probably nothing to be simplified. The inhomogeneous expressions might benefit from simplification, so I've added it back for them specifically.

**Comment by clinssen** [comment]
@pnbabu: thanks for the review! The latest commit should address all of the points raised. Could you re-review?

**Review by pnbabu — APPROVED** [review]
Looks good to me. Thanks!

**Comment by clinssen** [comment]
Many thanks!

**Pull request merged**